### PR TITLE
Ban Compilation.GetTypeByMetadataName

### LINKED ISSUE #2953: Ban the use of Compilation.GetTypeByMetadataName
We need to enable BannedApiAnalyzers on this repository, and ban the use of Compilation.GetTypeByMetadataName. Existing uses have already been updated, but we don't want more to slip in with future pull requests.

### PR DESCRIPTION
**Pull request by drewnoakes**
Fixes #2953

This is using the analyzer via NuGet.

### CODE REVIEW

File: src/Directory.Build.targets [hunk 1]
```diff
@@ -20,6 +20,11 @@
     <UpToDateCheckInput Include="$(MSBuildThisFileDirectory)..\eng\Analyzers_ShippingRules.ruleset" Condition="'$(CodeAnalysisRuleSet)' == '$(MSBuildThisFileDirectory)..\build\Analyzers_NonShippingRules.ruleset'" />
   </ItemGroup>
 
+  <ItemGroup>
+    <PackageReference Include="Microsoft.CodeAnalysis.BannedApiAnalyzers" Version="2.9.6" />
```
Comment (mavasani) [inline src/Directory.Build.targets]: Please move the package reference to eng\Versions.props

Comment (mavasani) [reply src/Directory.Build.targets]: Sorry, I meant version of the referenced package must be defined in Versions.props

Comment (sharwell) [reply src/Directory.Build.targets]: @mavasani I'll go ahead and fix this and the other issues that arose during build

Comment (mavasani) [reply src/Directory.Build.targets]: Thanks!

Comment (drewnoakes) [reply src/Directory.Build.targets]: Thanks @sharwell for picking this up while I slept :)

File: src/BannedSymbols.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+M:Microsoft.CodeAnalysis.Compilation.GetTypeByMetadataName(System.String); Use WellKnownTypeProvider instead
```
Comment (drewnoakes) [inline src/BannedSymbols.txt]: After seeing @sharwell's fixes, I think the diagnostic message here should probably have been

> Use GetOrCreateTypeByMetadataName with WellKnownTypeProvider instead

Comment (mavasani) [reply src/BannedSymbols.txt]: Agreed. I will fix it. Thanks @drewnoakes

**Pull request merged**